### PR TITLE
PHN5 - RGBD Camera

### LINKED ISSUE #5: Simulate RGBD Camera
**Description:**
- Photon will need a variety of sensors to complete the competition, so we add them in simulation.

**Goals:**
- [x] Add (stereo?) camera plugin - check if there is an RGBD plugin
- [ ] Add `stereo_image_proc` to produce:
  - [ ] Depth maps
  - [x] Point Clouds
  - [x] Rectified Images

**Acceptance Criteria:**
- [x] Relevant camera topics (e.g. images) are published
- [x] Camera outputs change when there is a change in the environment

**Notes:**
- It would be nice to keep track of what plugin arguments can be modified so we can more accurately simulate our sensors in the future.
- Depending on the camera we decide on for Photon, we may need to emulate some extra functionality.

### PR DESCRIPTION
**Pull request by psantasr**
### Description ###

Added the RGBD camera plug-in. Spawns in RViz and Gazebo. 

```bash
roslaunch description simulate.launch
roslaunch description view.launch
```

Test should see camera feed in RViz which aligns with its position in Gazebo

### Issues ###

Still unsure how to get it to look properly on gazebo.

### Notes ###

Looking for feedback to incorporate in the code and better it. Source of current code - Caffeine ZED Camera + MJ codes for the lidar plug-in in order to understand how to do a plug in 

### UPDATE ###

See my (@erickmu1) comment at the end of this PR. Also refer to Issue #5 where I show that the stereo images now look proper in RViz (as it mimics what is expected from Gazebo).

### CODE REVIEW

File: description/rviz/rgbd_camera.rviz [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    Property Tree Widget:
+      Expanded:
+        - /Global Options1
+        - /PointCloud21
+        - /Image2
+      Splitter Ratio: 0.5
+    Tree Height: 317
+  - Class: rviz/Selection
+    Name: Selection
+  - Class: rviz/Tool Properties
```
Comment (erickmu1) [inline description/rviz/rgbd_camera.rviz]: remove this rviz file and make respective changes in `view.rviz`

File: description/urdf/constants.xacro [hunk 2]
```diff
@@ -25,6 +25,28 @@
...
+    <xacro:property name="name" value="camera/camera_node"/>
+    <xacro:property name="always_on" value="true"/>
+    <xacro:property name="update_rate" value="0"/>
+    <xacro:property name="image_topicName" value="image_raw_color"/>
+    <xacro:property name="image_info_topicName" value="camera_info"/>
+    <xacro:property name="depth_topicName" value="depth/image_raw"/>
+    <xacro:property name="depth_info_topicName" value="depth/camera_info"/>
+    <xacro:property name="depth_points_topicName" value="depth/points"/>
+    <xacro:property name="frameName" value="zed_camera_link_optical"/>
     <!-- IMU -->
```
Comment (erickmu1) [inline description/urdf/constants.xacro]: new line (nice spacing)

File: description/urdf/cybertruck.urdf.xacro [hunk 3]
```diff
@@ -76,4 +76,28 @@
...
+                <box size="${camera_width} ${camera_length} ${camera_height}"/>
+            </geometry>
+        </collision>
+        <inertial>
+            <xacro:box_inertia mass="${camera_mass}" length="${camera_length}" width="${camera_width}" height="${camera_height}"/>
+            <mass value="${camera_mass}"/>
+        </inertial>
+    </link>
+    <joint name="camera_joint" type="fixed">
+        <origin xyz="${body_length/2 - camera_width} 0 ${camera_height / 2.0}" rpy="0 ${camera_pitch} 0"/>
```
Comment (erickmu1) [inline description/urdf/cybertruck.urdf.xacro]: replace `2` with `2.0`

**Review by erickmu1 — CHANGES_REQUESTED** [review]

File: description/launch/rgbd_camera.launch [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+            <remap from="points2"           to="point_cloud/cloud_registered"/>
+            <remap from="disparity"         to="disparity/disparity_registered"/>
+            <remap from="left/image_raw"    to="left/image_raw_color"/>
+            <remap from="right/image_raw"   to="right/image_raw_color"/>
+
+            <!-- <param name="speckle_range"     value="8"/> -->
+            <param name="speckle_size"      value="400"/>
+        </node>
+    </group>
+</launch>
```
Comment (EsteBran) [inline description/launch/rgbd_camera.launch:23]: add newline

File: description/launch/spawn.launch [hunk 5]
```diff
@@ -27,7 +27,11 @@
...
         <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
-    
+
+        <include file="$(find description)/launch/rgbd_camera.launch">
+            <arg name="camera_ns" value="camera/camera_node"/>
+        </include>
+
     </group>
     
 </launch>
```
Comment (EsteBran) [inline description/launch/spawn.launch:30]: add newline

File: description/launch/simulate.launch [hunk 6]
```diff
@@ -18,6 +17,11 @@
...
     </include>
 
+    <!-- Camera Plug-in -->
+    <include file="$(find description)/launch/rgbd_camera.launch">
+        <arg name="camera_ns" value="camera/camera_node"/>
+    </include>
+
     <!-- Spawn Robot -->
     <include file="$(find description)/launch/spawn_robot.launch" /> 
 </launch>
```
Comment (EsteBran) [inline description/launch/simulate.launch:19]: add newline

File: description/urdf/cybertruck.urdf.xacro (hunk 3 above)
Comment (EsteBran) [inline description/urdf/cybertruck.urdf.xacro]: Add newline

**Review by EsteBran — APPROVED** [review]
Reviewed, everything seems good, just a few newlines.

File: description/urdf/cybertruck.urdf.xacro [hunk 7]
```diff
@@ -76,4 +76,28 @@
         <origin xyz="0 0 ${-height_to_wheel}" rpy="0 0 0"/>
     </joint>
 
+    <!-- ZED Camera -->
+    <link name="camera_link">
+        <visual>
+            <origin xyz="0 0 0" rpy="0 0 0"/>
+            <geometry>
+            <box size="${camera_width} ${camera_length} ${camera_height}"/>
```
Comment (kishya) [inline description/urdf/cybertruck.urdf.xacro]: Formatting: Tab before <box

File: description/launch/simulate.launch [hunk 8]
```diff
@@ -18,6 +17,11 @@
 
     </include>
 
+    <!-- Camera Plug-in -->
```
Comment (kishya) [inline description/launch/simulate.launch]: the plugin is called twice: once from simulate.launch and once from spawn.launch.
spawn.launch is called in simulate.launch.

File: description/launch/spawn.launch [hunk 9]
```diff
@@ -27,7 +27,11 @@
         <!-- Setup TF Tree -->
         <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher"/>
         <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
-    
+
+        <include file="$(find description)/launch/rgbd_camera.launch">
```
Comment (kishya) [inline description/launch/spawn.launch]: Maybe add a comment explaining what this code block does

**Review by kishya — CHANGES_REQUESTED** [review]
A reminder to erase the hello.launch file

File: description/urdf/inertias.xacro [hunk 10]
```diff
@@ -21,5 +21,14 @@
         />
     </xacro:macro>
 
+    <!--Camera Inertia for the Box-->
+     <xacro:macro name="box_inertia" params="mass length width height">
+        <inertia
```
Comment (cuevasm2) [inline description/urdf/inertias.xacro:27]: Make the start `<inertia>` and end with `<\inertia>` instead of `/>`

**Review by erickmu1 — APPROVED** [review]
Please refer to the linked issue where I show a screenshot that shows the left/right stereo images align with Photon in RViz. At this point everything works as expected.

Note there is no depth images to view unless we add another node to perform `disparity --> depth` conversion. We did this in Caffeine using RTAB (which we also used for visual odom), so this can be deferred.

Also note that the `pointclouds` are off by default (since they are expensive to render)

**Pull request merged**